File: 1-100/3.longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chr2idx = {}
        nMaxLen = 0
        nChrBeg = -1

        for idx, c in enumerate(s):
            if c in chr2idx and nChrBeg < chr2idx[c]:
                nChrBeg = chr2idx[c]
            elif idx - nChrBeg > nMaxLen:
                nMaxLen = idx - nChrBeg
            chr2idx[c] = idx

        return nMaxLen

File: 1-100/3.longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(const string& s) {
        unordered_map<char, int> mapChrIdx;
        int nMaxLen = 0;
        int nChrBeg = -1;
        int nChrIdx = 0;
        for (const char& c : s)
        {
            if (mapChrIdx.count(c) == 1 && mapChrIdx[c] > nChrBeg)
            {
                nChrBeg = mapChrIdx[c];
            }
            else if (nChrIdx - nChrBeg > nMaxLen)
            {
                nMaxLen = nChrIdx - nChrBeg;
            }
            mapChrIdx[c] = nChrIdx;
            ++nChrIdx;
        }

        return nMaxLen;
    }
};